ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+  rescue Temporal::Error::ActivityCancelled
+    raise 'this activity is cancelled'
+  end
+end
+```
+
+Here is what it will look like from a workflow perspective:
+
+```ruby
+class InnerCancelWorkflow < Temporal::Workflow
+  def execute
+    # the block supplies an optional argument
+    async do |outer_promise|
+      async do
+        workflow.sleep(30)
+        outer_promise.cancel
+      end
+
+      # One of these blocking activities will end up getting cancelled 
+      #   unless they all finish within 30 seconds
+      workflow.execute_activity(LongRunningActivity)
+      workflow.execute_activity(LongRunningActivity)
+      workflow.execute_activity(LongRunningActivity)
+    end.await
+  end
+end
+```
+
+#### Alternative approaches
+
+- Overall using `Thread#raise` is considered bad practice because it might raise in the
+  middle of some process. We might want to consider opting in to this behaviour instead
+
+
+## Workflow cancellations
+
+To handle external workflow cancellations (requested by a client or another workflow) we will use a
+similar approach to activities. By default the cancellation will attempt to cancel all currently
+executing fibers. To protect critical functionality from cancellations a `workflow.shield` method
+can be used. In which case cancellation will raise after the block is executed:
+
+
+```ruby
+class MyWorkflow < Temporal::Activity
+  def execute
+    # Can get cancelled
+    workflow.execute_activity(NonCriticalActivity)
+
+    # This block will delay any cancellations until it is finished
+    async(shield: true) do
+      workflow.execute_activity(CriticalActivity)
+      workflow.execute_activity(OtherCriticalActivity)
+    end.await
+    
+    # Can get cancelled
+    workflow.execute_activity(AnotherNonCriticalActivity)
+  rescue Temporal::Error::WorkflowCancelled
+    ...
+  end
+end
+```
+
+
+## Register signal and query handlers
+
+`calculator_workflow.rb`
+
+```ruby
+class CalculatorWorkflow < Temporal::Workflow
+  # Both signal and query handlers can be defined as Symbols (when the name matches the
+  # method) or overriden when the name is different (or cannot be used as a method name)
+  signal :add
+  signal :subtract, name: 'my.signal.subtract'
+  
+  query :value # can use a getter here
+  query :get_previous_value, name: 'my.query.previous_value'
+
+  # It is preferable in this case to initialize the values before the execution to
+  # support starting a workflow with a signal
+  def initialize
+    @value = 0
+    @previous_value = 0
+  end
+
+  def execute
+    async.await_for { @value >= 42 }
+  end
+
+  private
+  
+  attr_reader :value
+
+  def add(val)
+    @previous_value = @value
+    @value += val
+  end
+
+  def subtract(val)
+    @previous_value = @value
+    @value -= val
+  end
+
+  def get_previous_value
+    @previous_value
+  end
+end
+```
+
+#### Alternative approach
+
+To provide extra flexibility in handling signals and queries (especially from external classes and
```

Comment:
I'm in favor of the the pattern above instead of the one below if possible. I've been able to compare workflow authors' use of the above approach in Java with the below in temporal-ruby since we support both equally at Stripe. I've seen more confusion leading to code defects with the callback-based approach. Mainly authors don't know when they should register their signal handler, and by not understanding that the code is not truly multithreaded. The above approach solves the first problem well, by taking way this unnecessary control. I've yet to find a case where registering a callback later is necessary. The second problem remains but seems to be less confusing because receipt of signals is less intertwined with their "main" code.

This got bad enough that we ended up eventually banning the below approach, abstracting over it in favor of an approach similar to that of the Go SDK. That is another option to consider here potentially as well if you do want to offer an inline approach. It ends up looking something like:

```
  queue = workflow.get_signal_queue(:foo)
  signal = queue.receive # blocks indefinitely
  signal = queue.receive(timeout: 60) # blocking with timeout, raises on timeout
  maybe_signal = queue.receive_now # returns nil if there is no signal available
  queue.empty? # true if there are no signals
```

However, there's not an obvious way to inline queries like this.

Reply:
I believe the full team agrees with you here. While our examples and suggestions will be the explicit methods over callbacks, we still often need the callbacks to exist for dynamic registration based on input. So I think both being present with strong emphasis on the method-based approach is ideal (and what Python does).

Reply:
The main purpose for the alternative approach here is to provide the flexibility of extending the behaviour:

```ruby
class MyWorkflow < Temporal::Workflow
  def execute
    ExecutionTimeQueryHandler.track(workflow) do
       ...
    end
  end
end

class ExecutionTimeQueryHandler
  def self.track(workflow, &block)
    @start_time = workflow.now
    workflow.on_signal(:get_time) { workflow.now - @time }
    block.call
  end
end
```

But this is more of an "advanced usage" rather than the default approach we'd recommend